5594 fix portspace detection

hessenfarmer opened the pull request:
fixes #5594 
fixes #5155 

- rocks or ruins will block a portspace from detection
- trees will not block detection
- now all trees including immature ones will be removed at construction
- ownership and availability is rechecked periodically and before constructing a port (might be conquered from detection to construction
- a flag at the flag field and some roads around the 4 needed spaces are allowed (build the expedition port on own territory)

hessenfarmer committed:
fix portspace detection
- rocks or ruins will block a portspace from detection
- trees will not block detection

frankystone commented:
Thanks for a more detailed explanation.

Works as described :-)

tothxa commented:
I had already built this for testing, just hadn't got around to test it. :)

The change looks OK, but I've found a problem in testing. The port space next to the rocks in the following image isn't found, even though it is shown as buildable in the editor:

![portspace_editor](https://user-images.githubusercontent.com/80712255/200201547-3e102a34-6760-464b-8b02-2511130fa7fd.png)

(yes I wanted to push it to the limit by also testing whether clearing nearby area still works ;)

If I remove the rock to the East of the portspace's flag position, then it is found. (and the other rocks are cleared by the expedition when starting to build the port)

I'm not sure though if we could limit the check to the 4 fields of the portspace plus the flag in `can_build_port_here()`. I'm inclined to think we shouldn't.

BTW I've found a bug with the clearing code: Only fully grown trees are removed:

![growing_tree_not_removed](https://user-images.githubusercontent.com/80712255/200202487-9b005416-151f-49fd-b096-f6e96311d1d5.png)

Should I open a new bug for this or can it be fixed in this PR?

Here's the map I modified for testing: 
[portspace_test_map.zip](https://github.com/widelands/widelands/files/9947128/portspace_test_map.zip)

hessenfarmer commented:
Will try to fix both tonight:  
for the clearance issue we currently use the attrib "tree" which only full grown trees have. Will use has_terrain affinity there as well.     
for the other issue we scan each field in the radius of 1 for the reason to be far enough from any border. so we might need to split the check for the owner of a field and the check for being occupied by something solid.

hessenfarmer committed:
fix clearance bug

hessenfarmer committed:
seperate immovable check from owner check
some optimizations

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 5594_portspace_detection

Commit:
'src/logic/map_objects/tribes/ship.cc' was automatically formatted.

tothxa reviewed (commented):
Tested the changes, and now everything worked as intended. I think I've found a copy-paste mistake in the code, see inline. Testing its effect is hard, because it's a special corner case if I'm not mistaken. (only the base coordinate is checked for ownership, not the surrounding area of all fields occupied by the port)

hessenfarmer committed:
fix copy paste error

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

hessenfarmer commented:
thanks for spotting my mistakes
Hopefully this will fix the testsuite as well

tothxa reviewed (approved):
Thank you. I think it's fine now.

I've found yet another old bug though while testing portspaces near borders that could be fixed here if you don't mind yet another round: `exp_construct_port()` should call `can_build_port_here()` too as a check before placing port. If another player conquers the area while the ship is waiting at a portspace, it is now possible to build a port and take the land away from the other player. Of course if you move the ship away and bring it back it doesn't find the portspace any more. [savegame for illustration](https://github.com/widelands/widelands/files/9961702/portspace_found_conquered.zip)

Ideally ships waiting at a portspace should also regularly re-check the availability of that portspace.

hessenfarmer commented:
Now that we touched this we will surely fix every bug that arises in testing. no matter how many rounds it takes.   
thanks for finding this very old cornercase bug

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 5594_portspace_detection

tothxa commented:
I've found another old bug report concerning portspace detection: #5155

Unfortunately the savegame can't be loaded even by v1.1, so I'll try to recreate the situation and check with this branch.

edit: done, bug exists with this branch too, new savegame for testing was uploaded to original bug report

hessenfarmer committed:
recheck ownership before constructing expedition port
optimized ownership check

hessenfarmer committed:
fix 5155

tothxa reviewed (commented):
Thank you for the fixes. See comment inline.

tothxa commented:
Sorry, there is a regression in the fix. The neighbouring fields of the 4 main fields (but not the other neighbours of the flag) have to be checked for stones to only find portspaces that are also buildable from land.

hessenfarmer commented:
> Sorry, there is a regression in the fix. The neighbouring fields of the 4 main fields (but not the other neighbours of the flag) have to be checked for stones to only find portspaces that are also buildable from land.

this would make a fix for #5155 at least very hard. Furthermore why should a port not be buildable from land if one of the neighbours is occupied by a rock

tothxa commented:
> Furthermore why should a port not be buildable from land if one of the neighbours is occupied by a rock

Well, this is what #5594 was about. :) Because the port space is shown as blocked in the editor and in the game as well if it is conquered from land. Spreading of trees complicates the issue, but at least in the absence of trees, the expedition check should give the same result as the normal buildspace check. (actually I even think that if possible, the tree issue should be solved by not allowing automatic seeding of trees near portspaces instead of special casing them here, to not mislead map designers)

hessenfarmer commented:
ok I see the point. I forgot that rocks are big.  Will try to fix

tothxa commented:
I think we need these checks (size 0 and trees count as nothing, unless seeding is solved, in which case we just have to check buildcap and ownership):

 - nothing is allowed on coords, ln, tln, trn
 - only own flag is allowed at flagpos = brn
 - only own road is allowed at bln and rn
 - size 1 except flag is allowed at brn, bln and rn of flagpos
 - only own road or own flag is allowed at trn+rn, trn+trn, trn+tln, tln+tln, ln+tln, ln+ln, ln+bln

hessenfarmer commented:
yep, will try to call / copy the logic of the editor, or the game. Just need to find it. 
I am still struggling with the periodic check. Need to follow a complete new idea. maybe tonight.

hessenfarmer committed:
add periodic check for port space availability

hessenfarmer committed:
fix regression

Commit:
'src/logic/map_objects/tribes/ship.cc' was automatically formatted.

tothxa reviewed (commented):
Thank you. I think everything works now. Just some nitpicking inline. :)

For the record, there's a corner case of somewhat inconcistent behaviour with owned land because own flags aren't ignored now:

 - if only trees are present, then the portspace is found
 - if trees and otherwise compatible roads and flags are present then the portspace is not found
 - (of course if the portspace is clear and only compatible roads and flags are present, then the first check ensures that it is found)

But as this is already conquered land, it shouldn't matter much as #5155 is fixed by the first check.

hessenfarmer committed:
update comment

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

hessenfarmer committed:
explain better that trees are allowed on a portspace

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

hessenfarmer committed:
Merge branch 'master' into 5594_portspace_detection

hessenfarmer committed:
fix test suite regression

hessenfarmer committed:
logging possible cornercase

Commit:
'src/logic/map_objects/tribes/ship.cc' was automatically formatted.

tothxa commented:
Re testsuite regression: Shouldn't we add tests for this?

hessenfarmer committed:
fix typo in log message

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

hessenfarmer commented:
> Re testsuite regression: Shouldn't we add tests for this?       
   
Well we have the tests and they failed so they are good enough imho. the problem was that we need to check for ownership first, because else the normal port build check fails. What else would you have checked. 

Edit: ok now I understand. Do you want us to check for trees allowed - rocks not?

hessenfarmer committed:
pulled out a common function

Commit:
'src/logic/map_objects/tribes/ship.cc' was automatically formatted.

hessenfarmer committed:
added tests

hessenfarmer committed:
Merge branch '5594_portspace_detection' of https://github.com/widelands/widelands into 5594_portspace_detection

tothxa reviewed (commented):
Thank you very much. :) (I thought `recheck_portspace()` would suffice for function name. :)

Some requests inline for extending test coverage. Furthermore, please also add a test for #5155 by placing nearest possible road with flags around the portspace. Maybe again test eliminating portspace afterwards by placing a road or some building inside.

hessenfarmer committed:
try to make clang tidy happy

hessenfarmer committed:
readability improvement

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

hessenfarmer committed:
Merge branch 'master' of https://github.com/widelands/widelands into 5594_portspace_detection

Commit:
'src/logic/map_objects/tribes/ship.cc' was automatically formatted.

hessenfarmer committed:
fix message to be detectable in testsuite

hessenfarmer committed:
added some tests

hessenfarmer committed:
Merge branch '5594_portspace_detection' of https://github.com/widelands/widelands into 5594_portspace_detection

hessenfarmer commented:
Hopefully I am done now ;-)

tothxa commented:
> Hopefully I am done now ;-)

I'll need some time to go over the tests in detail. :( I hoped I'd be able to watch them progress, but I'm afraid that's not possible. Checking them blindly will be harder.

hessenfarmer committed:
typo correction

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

hessenfarmer committed:
fix indexing issue

Co-authored-by: Tóth András <80712255+tothxa@users.noreply.github.com>

hessenfarmer committed:
fix comment

hessenfarmer commented:
@tothxa  any further comments

tothxa commented:
I've modified the test map and one of the test cases to test the port space surrounded by rocks at the nearest possible positions, and added some size small immovables where the buildcap checks allow them: 
[port_space_wmf.zip](https://github.com/widelands/widelands/files/10044710/port_space_wmf.zip)

This however fails, but I think it shouldn't. Here's the new portspace in the editor without trees:

![portspace](https://user-images.githubusercontent.com/80712255/202799693-d16b7116-c85d-446c-a52f-42f8fa18b929.png)

So adding only trees shouldn't make it fail. I'm not sure it's only the artifacts that make it fail though, because it still fails if I remove everything.

tothxa commented:
I've finally looked through all tests. (even figured out that I can actually watch them work :) They're OK, I think only my above test case is needed in addition.

I've modified `wait_for_message()` to abort on wrong message to not depend on time-outs. Not sure if it's safe though, but all tests (except for the above failure of course) passed for me with this change. (I also wonder why it has several long delays, but I didn't touch them.) I also added progress messages to the tests which have multiple steps:

[port_space_test_wmf.zip](https://github.com/widelands/widelands/files/10045700/port_space_test_wmf.zip)

tothxa commented:
Here's a patch that makes detection work the same as in the editor, except for trees (I checked that medium isn't allowed on the Northern/Western 2nd neighbours): 
[size_small_fix_patch.txt](https://github.com/widelands/widelands/files/10046540/size_small_fix_patch.txt) (edit: fixed East->West in comment)

My test case still doesn't find it, but I can't see why. On my big test map the same immovable arrangement works, so it must be some interference from the other port or the owned land, but I couldn't find in the code how that could affect the result.

Sorry for all the trouble I make.

hessenfarmer commented:
> Here's a patch that makes detection work the same as in the editor, except for trees (I checked that medium isn't allowed on the Northern/Western 2nd neighbours): [size_small_fix_patch.txt](https://github.com/widelands/widelands/files/10046540/size_small_fix_patch.txt) (edit: fixed East->West in comment)
> 
> My test case still doesn't find it, but I can't see why. On my big test map the same immovable arrangement works, so it must be some interference from the other port or the owned land, but I couldn't find in the code how that could affect the result.
> 
> Sorry for all the trouble I make.

no need to apologize: testing my buggy code thoroughly isn't making trouble it is adding quality. So thanks for doing so.   
I added a patch quite similar to yours and ran into the same issue that it is not finding it. So I started to add some debug code but did not succeed yet

tothxa commented:
It's the portdock check. It isn't allowed on the border. Here's the fixed test case: 
[port_space_test_wmf.zip](https://github.com/widelands/widelands/files/10046571/port_space_test_wmf.zip)

edit: OTOH border with neutral shouldn't fail. At first sight it may seem as a narrow corner case, but actually it may make a neighbouring island unconquerable. I'm afraid though that checking whether other size is neutral or other player wouldn't be trivial.

hessenfarmer commented:
> It's the portdock check. It isn't allowed on the border. Here's the fixed test case: [port_space_test_wmf.zip](https://github.com/widelands/widelands/files/10046571/port_space_test_wmf.zip)
> 
> edit: OTOH border with neutral shouldn't fail. At first sight it may seem as a narrow corner case, but actually it may make a neighbouring island unconquerable. I'm afraid though that checking whether other size is neutral or other player wouldn't be trivial.

I think we can safely use the force true in ship.cc line 441. will test that. and report back

tothxa committed:
Fix handling of small immovables and add test case with closest allowed immovables (#5663)

hessenfarmer committed:
expedition first portdockdetection adopted to detect all theoretical port spaces.

hessenfarmer commented:
ok using the force option of find_portdock shpul dreally deliver all theoretical places as intended. 
Should be save as we check them later on . at least it worked locally lets see what the test suite says.

tothxa commented:
> ok using the force option of find_portdock shpul dreally deliver all theoretical places as intended. Should be save as we check them later on . at least it worked locally lets see what the test suite says.

Indeed. I think it's fine now if all tests pass.

tothxa reviewed (approved):
Thank you. :)

hessenfarmer commented:
> Thank you. :)

Thank you for such good testing and inputs
